---
title: "Bamboo mesocosms experiment"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
 rmarkdown::html_document:
    theme: cerulean
    toc: yes
    toc_float: yes
    highlight: haddock
    number_sections: true
editor_options: 
  chunk_output_type: console
---

# Setup

## Libs

```{r}
#install.packages("remotes")
#install.packages('TMB',source=TRUE) #didn't work on home laptop
#remotes::install_github("glmmTMB/glmmTMB/glmmTMB")
library("glmmTMB")
#install.packages("BiocManager")
#BiocManager::install("phyloseq")
library("phyloseq")
library("dplyr")
# #install.packages("reshape")
library("reshape")
# library("ggplot2")
# #install.packages("bbmle")
library("bbmle")
# #install.packages("car")
# library("car")
library("DHARMa")
library("plyr")
library("ggplot2")
```

## Re-read data

```{r}
setwd("~/nicolagk@hawaii.edu - Google Drive/My Drive/Bamboo_mesos/Bamboo_mesos/04.mixed_models/")

#ps.clean <- readRDS("../01.process_asvs/reusum23.ps.lulu.clean.rds")
ps.clean <- readRDS("../01.process_asvs/bamboomesos_ps.lulu.clean.decontam.rds")
#ps.clean <- readRDS("../01.process_asvs/ps.all.clean.100.rds")
ps.clean #514 taxa, 169 samples

ps.exp <- subset_samples(ps.clean,Mesocosm_type=="Experiment"&Microbe_treatment!="AHK"&Exp_day!=20&Microbe_treatment!="ECO"&Microbe_treatment!="ALL")
ps.exp #60 samples
##including ECO & ALL to compare
ps.exp.all <- subset_samples(ps.clean,Mesocosm_type=="Experiment"&Microbe_treatment!="AHK"&Exp_day!=20)
ps.exp.all #100 samples

##water samples
ps.exp.wtr <- subset_samples(ps.exp,Sample_type=="Water")
ps.exp.wtr #30 samples
##plus eco & all
ps.exp.all.wtr <- subset_samples(ps.exp.all,Sample_type=="Water")
ps.exp.all.wtr #50 samples

##larvae samples
ps.exp.lar <- subset_samples(ps.exp,Sample_type=="Larvae")
ps.exp.lar #30 samples
##plus eco & all
ps.exp.all.lar <- subset_samples(ps.exp.all,Sample_type=="Larvae")
ps.exp.all.lar #50 samples
```

# Data formatting

## Trim, thoroughly

Ran once then saved

```{r}
##6 samples out of 50 = 12%
##6 samples out of 30 = 20%

##water
ps.exp.wtr.trim <- filter_taxa(ps.exp.wtr, function(x) sum(x > 1) > (0.20*length(x)), TRUE)
ps.exp.wtr.trim ##62 taxa
tail(sort(sample_sums(ps.exp.wtr.trim),decreasing=T))
##with all and eco
ps.exp.all.wtr.trim <- filter_taxa(ps.exp.all.wtr, function(x) sum(x > 1) > (0.12*length(x)), TRUE)
ps.exp.all.wtr.trim #69 taxa

##larvae
#ps.exp.lar.trim <- filter_taxa(ps.exp.lar, function(x) sum(x > 1) > (0.20*length(x)), TRUE)
#ps.exp.lar.trim #44 taxa 
##with all and eco
ps.exp.all.lar.trim <- filter_taxa(ps.exp.all.lar, function(x) sum(x > 1) > (0.12*length(x)), TRUE)
#ps.exp.all.lar.trim <- filter_taxa(ps.exp.all.lar, function(x) sum(x > 1) > (0.15*length(x)), TRUE)

ps.exp.all.lar.trim #54 taxa

tail(sort(sample_sums(ps.exp.lar.trim),decreasing=T))

##proportion of zeroes
##water
otu.exp.wtr <- data.frame(ps.exp.wtr@otu_table)
sum(otu.exp.wtr == 0) / (ncol(otu.exp.wtr) * nrow(otu.exp.wtr)) * 100
#93.19066%
 
otu.exp.wtr.trim <- data.frame(ps.exp.wtr.trim@otu_table)
sum(otu.exp.wtr.trim == 0) / (ncol(otu.exp.wtr.trim) * nrow(otu.exp.wtr.trim)) * 100
#54.94624%

##water - with all
otu.exp.all.wtr <- data.frame(ps.exp.all.wtr@otu_table)
sum(otu.exp.all.wtr == 0) / (ncol(otu.exp.all.wtr) * nrow(otu.exp.all.wtr)) * 100
#94.27237%
 
otu.exp.all.wtr.trim <- data.frame(ps.exp.all.wtr.trim@otu_table)
sum(otu.exp.all.wtr.trim == 0) / (ncol(otu.exp.all.wtr.trim) * nrow(otu.exp.all.wtr.trim)) * 100
#64.11594%

##save if ready
#saveRDS(ps.exp.wtr.trim,file="ps.exp.wtr.trim.rds")
#saveRDS(ps.exp.all.wtr.trim,file="ps.exp.all.wtr.trim.rds")

##larvae
otu.exp.lar <- data.frame(ps.exp.lar@otu_table)
sum(otu.exp.lar == 0) / (ncol(otu.exp.lar) * nrow(otu.exp.lar)) * 100
#95.32425%
 
otu.exp.lar.trim <- data.frame(ps.exp.lar.trim@otu_table)
sum(otu.exp.lar.trim == 0) / (ncol(otu.exp.lar.trim) * nrow(otu.exp.lar.trim)) * 100
#61.81818%

##larvae - with all
otu.exp.all.lar <- data.frame(ps.exp.all.lar@otu_table)
sum(otu.exp.all.lar == 0) / (ncol(otu.exp.all.lar) * nrow(otu.exp.all.lar)) * 100
#95.96498%
 
otu.exp.all.lar.trim <- data.frame(ps.exp.all.lar.trim@otu_table)
sum(otu.exp.all.lar.trim == 0) / (ncol(otu.exp.all.lar.trim) * nrow(otu.exp.all.lar.trim)) * 100
#70.96296%

##save if ready
#saveRDS(ps.exp.lar.trim,file="ps.exp.lar.trim.rds")
#saveRDS(ps.exp.all.lar.trim,file="ps.exp.all.lar.trim.rds")
```

### Which ASVs to subset from fasta

Didn't do yet

```{r}
otus.lar.trim <- colnames(otu.exp.lar.trim)
otus.wtr.trim <- colnames(otu.exp.wtr.trim)

#write.table(otus.lar.trim, file="otus.larvae.trim.txt", append = FALSE, sep = "/n", row.names = FALSE, col.names = FALSE,quote=FALSE)
#write.table(otus.wtr.trim, file="otus.water.trim.txt", append = FALSE, sep = "/n", row.names = FALSE, col.names = FALSE,quote=FALSE)
```

## More formatting - long form{.tabset}

### Just EMO, MMO, and MEM 

```{r}
otu.wtr <- data.frame(ps.exp.wtr.trim@otu_table)
otu.lar <- data.frame(ps.exp.lar.trim@otu_table)

#total read count colums
otu.wtr$sum <- rowSums(otu.wtr)
otu.lar$sum <- rowSums(otu.lar)

otu.wtr$Short_label <- row.names(otu.wtr)
otu.lar$Short_label <- row.names(otu.lar)

#bring in the metadata again
samdf.wtr <- data.frame(ps.exp.wtr.trim@sam_data)
samdf.lar <- data.frame(ps.exp.lar.trim@sam_data)

##add extra metadata
#meso.data <- read.csv("../bamboomesos_metadata.csv")

#samdf.wtr.more <- merge(samdf.wtr,meso.data)
#samdf.lar.more <- merge(samdf.lar,meso.data)

#samdf.wtr$food.microbes <- paste0(samdf.wtr$Food_type,"_",samdf.wtr$Microbe_treatment)
#samdf.lar$food.microbes <- paste0(samdf.lar$Food_type,"_",samdf.lar$Microbe_treatment)
# 
# #join the data sets by sample name
df.water <- plyr::join(otu.wtr, samdf.wtr, by="Short_label", type="left")
df.larv <- plyr::join(otu.lar, samdf.lar, by="Short_label", type="left")
colnames(df.water)

##getting rid of some irrelevant metadat
colnames(df.water)
df.water1 <- df.water %>%
  dplyr::select(-Short_label,-Longer_name,-Sample_type,-Num_larvae,-Date_collected,-Exp_day,-Mesocosm_treatment,-Mesocosm_type,-Treatment_notes,-Notes,-Raw_reads,-Num_larvae_d00,-Num_larvae_d08,-Num_larvae_d34,-Num_larvae_d40,-Total_pupae,-Proportion_pupated,-Biomass_day30,-lib_size_clean,-is.neg)
colnames(df.water1)

df.water.long <- melt(df.water1, id.vars=c("sum", "Mesocosm_id", "Microbe_treatment","Food_type", "Biomass_day40"))

df.water.long$rowid <- 1:nrow(df.water.long)

df.larv1 <- df.larv %>%
  dplyr::select(-Short_label,-Longer_name,-Sample_type,-Num_larvae,-Date_collected,-Exp_day,-Mesocosm_treatment,-Mesocosm_type,-Treatment_notes,-Notes,-Raw_reads,-Num_larvae_d00,-Num_larvae_d08,-Num_larvae_d34,-Num_larvae_d40,-Total_pupae,-Proportion_pupated,-Biomass_day30,-lib_size_clean,-is.neg)
colnames(df.larv1)

#df.larv2 <- df.larv1[df.larv1$sum!=0,]
#df.larv2$sum==0

df.larv.long <- melt(df.larv1, id.vars=c("sum", "Mesocosm_id", "Microbe_treatment","Food_type", "Biomass_day40"))

df.larv.long$rowid <- 1:nrow(df.larv.long)

##save
#saveRDS(df.water.long,file="df.water.long.rds")
#saveRDS(df.larv.long,file="df.larv.long.rds")
```

### All of 'em

```{r}
otu.all.wtr <- data.frame(ps.exp.all.wtr.trim@otu_table)
otu.all.lar <- data.frame(ps.exp.all.lar.trim@otu_table)

#total read count colums
otu.all.wtr$sum <- rowSums(otu.all.wtr)
otu.all.lar$sum <- rowSums(otu.all.lar)

otu.all.wtr$Short_label <- row.names(otu.all.wtr)
otu.all.lar$Short_label <- row.names(otu.all.lar)

#bring in the metadata again
samdf.all.wtr <- data.frame(ps.exp.all.wtr.trim@sam_data)
samdf.all.lar <- data.frame(ps.exp.all.lar.trim@sam_data)

##add extra metadata
#meso.data <- read.csv("../bamboomesos_metadata.csv")

#samdf.wtr.more <- merge(samdf.wtr,meso.data)
#samdf.lar.more <- merge(samdf.lar,meso.data)

#samdf.wtr$food.microbes <- paste0(samdf.wtr$Food_type,"_",samdf.wtr$Microbe_treatment)
#samdf.lar$food.microbes <- paste0(samdf.lar$Food_type,"_",samdf.lar$Microbe_treatment)
# 
# #join the data sets by sample name
df.all.water <- plyr::join(otu.all.wtr, samdf.all.wtr, by="Short_label", type="left")
df.all.larv <- plyr::join(otu.all.lar, samdf.all.lar, by="Short_label", type="left")
colnames(df.all.water)

##getting rid of some irrelevant metadat
df.all.water1 <- df.all.water %>%
  dplyr::select(-Short_label,-Longer_name,-Sample_type,-Num_larvae,-Date_collected,-Exp_day,-Mesocosm_treatment,-Mesocosm_type,-Treatment_notes,-Notes,-Raw_reads,-Num_larvae_d00,-Num_larvae_d08,-Num_larvae_d34,-Num_larvae_d40,-Total_pupae,-Proportion_pupated,-Biomass_day30,-lib_size_clean,-is.neg)
colnames(df.all.water1)

df.all.water.long <- melt(df.all.water1, id.vars=c("sum", "Mesocosm_id", "Microbe_treatment","Food_type", "Biomass_day40"))

df.all.water.long$rowid <- 1:nrow(df.all.water.long)

df.all.larv1 <- df.all.larv %>%
  dplyr::select(-Short_label,-Longer_name,-Sample_type,-Num_larvae,-Date_collected,-Exp_day,-Mesocosm_treatment,-Mesocosm_type,-Treatment_notes,-Notes,-Raw_reads,-Num_larvae_d00,-Num_larvae_d08,-Num_larvae_d34,-Num_larvae_d40,-Total_pupae,-Proportion_pupated,-Biomass_day30,-lib_size_clean,-is.neg)
colnames(df.all.larv1)

#df.all.larv2 <- df.all.larv1[df.all.larv1$sum!=0,]
#df.all.larv2$sum==0

df.all.larv.long <- melt(df.all.larv1, id.vars=c("sum", "Mesocosm_id", "Microbe_treatment","Food_type", "Biomass_day40"))

df.all.larv.long$rowid <- 1:nrow(df.all.larv.long)

##save
#saveRDS(df.all.water.long,file="df.all.water.long.rds")
#saveRDS(df.all.larv.long,file="df.all.larv.long.rds")
```

# Running the statz{.tabset}

Can start from here

```{r}
setwd("~/nicolagk@hawaii.edu - Google Drive/My Drive/Bamboo_mesos/Bamboo_mesos/04.mixed_models")

df.water.long <- readRDS("df.water.long.rds")
df.larv.long <- readRDS("df.larv.long.rds")

df.all.water.long <- readRDS("df.all.water.long.rds")
df.all.larv.long <- readRDS("df.all.larv.long.rds")
```

## Water

```{r}
str(df.water.long)
df.water.long$rowid <- as.factor(df.water.long$rowid)
df.water.long$Microbe_treatment <- as.factor(df.water.long$Microbe_treatment)
df.water.long$Food_type <- as.factor(df.water.long$Food_type)
#df.water.long$food.microbes <- as.factor(df.water.long$food.microbes)
df.water.long$Mesocosm_id <- as.factor(df.water.long$Mesocosm_id)
str(df.water.long)

##full model
water.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~variable+Microbe_treatment+Food_type,data=df.water.long)
summary(water.mod.nb1)

water.mod.nb2 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom2,dispformula=~variable+Microbe_treatment+Food_type,data=df.water.long)
summary(water.mod.nb2)

AICtab(water.mod.nb1,water.mod.nb2)

water.mod.nb1.resid <- simulateResiduals(fittedModel = water.mod.nb1, plot = T)
water.mod.nb2.resid <- simulateResiduals(fittedModel = water.mod.nb2, plot = T)

##dispersal formula checks
water.mod.nb1.disnovar <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~Microbe_treatment+Food_type,data=df.water.long)

anova(water.mod.nb1.disnovar,water.mod.nb1) #sig***

##removing 3 way interaction
water.mod.nb1.nofm <- update(water.mod.nb1,.~.-(1|variable:Microbe_treatment:Food_type))
anova(water.mod.nb1,water.mod.nb1.nofm)
##super sig 0.000242

water.mod.nb2.nofm <- update(water.mod.nb2,.~.-(1|variable:Microbe_treatment:Food_type))
anova(water.mod.nb2,water.mod.nb2.nofm)
##also super sig

###save conditional modes
water.mod.ranef <- as.data.frame(ranef(water.mod.nb1))
#write.csv(water.mod.ranef, "water.mod.ranef.csv")
#saveRDS(water.mod.nb1,file="water.mod.nb1.rds")
```

### Nbinom1 vs. nbinom2 things

Following this wonderful resource, on page 16

Ben Bolker, Mollie Brooks, Beth Gardner, Cleridy Lennert, Mihoko Minami, October 23, 2012, Owls example: a zero-inflated, generalized linear mixed model for count data. https://groups.nceas.ucsb.edu/non-linear-modeling/projects/owls/WRITEUP/owls.pdf

```{r}
mvtab <- ddply(df.water.long,
               .(variable:Microbe_treatment:Food_type),
               summarise,
               callmean=mean(value),
               callvar=var(value))

q1 <- qplot(callmean,callvar,data=mvtab)
print(q1+
        ## linear (quasi-Poisson/NB1) fit
        geom_smooth(method="lm",formula=y~x-1,colour="pink")+
        ## smooth (loess)
        geom_smooth(colour="red")+
        ## semi-quadratic (NB2/LNP)
        geom_smooth(method="lm",formula=y~I(x^2)+offset(x)-1,colour="purple")+
        ## Poisson (v=m)
        geom_abline(a=0,b=1,lty=2))
```

## Larvae

```{r}
str(df.larv.long)
df.larv.long$rowid <- as.factor(df.larv.long$rowid)
df.larv.long$Microbe_treatment <- as.factor(df.larv.long$Microbe_treatment)
df.larv.long$Food_type <- as.factor(df.larv.long$Food_type)
#df.larv.long$food.microbes <- as.factor(df.larv.long$food.microbes)
df.larv.long$Mesocosm_id <- as.factor(df.larv.long$Mesocosm_id)
str(df.larv.long)

##binomial
larv.mod.bin <- glmmTMB(cbind(value, sum-value)~Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type)+(1|rowid),family=binomial,data=df.larv.long)
summary(larv.mod.bin)

##full model
larv.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),disp=~variable+Food_type+Microbe_treatment,family=nbinom1,data=df.larv.long)
summary(larv.mod.nb1)
#saveRDS(larv.mod.nb1,file="larv.mod.nb1.rds")

larv.mod.nb2 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),disp=~variable+Food_type+Microbe_treatment,family=nbinom2,data=df.larv.long)
summary(larv.mod.nb2)

AICtab(larv.mod.nb1,larv.mod.nb2,larv.mod.bin)

larv.mod.bin.resid <- simulateResiduals(fittedModel = larv.mod.bin, plot = T)
larv.mod.nb1.resid <- simulateResiduals(fittedModel = larv.mod.nb1, plot = T)
larv.mod.nb2.resid <- simulateResiduals(fittedModel = larv.mod.nb2, plot = T)

##dispersal formula checks
larv.mod.nb1.disnovar <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~Microbe_treatment+Food_type,data=df.larv.long)

anova(larv.mod.nb1.disnovar,larv.mod.nb1) #sig***

##removing 3 way interaction
larv.mod.nb1.nofm <- update(larv.mod.nb1,.~.-(1|variable:Microbe_treatment:Food_type))
summary(larv.mod.nb1.nofm)
anova(larv.mod.nb1,larv.mod.nb1.nofm) #sig **

# ##removing just food
# larv.mod.nb1.nofood <- update(larv.mod.nb1,.~.-(1|variable:Food_type))
# anova(larv.mod.nb1,larv.mod.nb1.nofood) #super sig***
# 
# ##removing just microbes
# larv.mod.nb1.nomicr <- update(larv.mod.nb1,.~.-(1|variable:Microbe_treatment))
# anova(larv.mod.nb1,larv.mod.nb1.nomicr) #super sig***

##checking out nbinom2
larv.mod.nb2.nofm <- update(larv.mod.nb2,.~.-(1|variable:Microbe_treatment:Food_type))
anova(larv.mod.nb2,larv.mod.nb2.nofm)
##sig**

##checking out binomial
larv.mod.bin.nofm <- update(larv.mod.bin,.~.-(1|variable:Microbe_treatment:Food_type))
summary(larv.mod.bin.nofm)
anova(larv.mod.bin,larv.mod.bin.nofm) #sig***

###save conditional modes
larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb1))
#write.csv(larv.mod.ranef, "larv.mod.ranef.csv")

# library("ggstats")
# ggcoef_multicomponents(larv.mod.nb1)
```

### Nbinom1 vs. nbinom2 things

```{r}
##note: some errors/warnings if the treatments aren't factors for some reason
mvtab <- ddply(df.larv.long,
               .(variable:Microbe_treatment:Food_type),
               summarise,
               callmean=mean(value),
               callvar=var(value))

q1 <- qplot(callmean,callvar,data=mvtab)
print(q1+
        ## linear (quasi-Poisson/NB1) fit
        geom_smooth(method="lm",formula=y~x-1,colour="pink")+
        ## smooth (loess)
        geom_smooth(colour="red")+
        ## semi-quadratic (NB2/LNP)
        geom_smooth(method="lm",formula=y~I(x^2)+offset(x)-1,colour="purple")+
        ## Poisson (v=m)
        geom_abline(a=0,b=1,lty=2))#+
        #ylim(0,1.75e8)+
        #xlim(0,20000)
```

## Water - all treatments

```{r}
str(df.all.water.long)
df.all.water.long$rowid <- as.factor(df.all.water.long$rowid)
df.all.water.long$Microbe_treatment <- as.factor(df.all.water.long$Microbe_treatment)
df.all.water.long$Food_type <- as.factor(df.all.water.long$Food_type)
#df.all.water.long$food.microbes <- as.factor(df.all.water.long$food.microbes)
df.all.water.long$Mesocosm_id <- as.factor(df.all.water.long$Mesocosm_id)
str(df.all.water.long)

##full model
water.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~variable+Microbe_treatment+Food_type,data=df.all.water.long)
summary(water.mod.nb1)

water.mod.nb2 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom2,dispformula=~variable+Microbe_treatment+Food_type,data=df.all.water.long)
summary(water.mod.nb2)

AICtab(water.mod.nb1,water.mod.nb2)

water.mod.nb1.resid <- simulateResiduals(fittedModel = water.mod.nb1, plot = T)
water.mod.nb2.resid <- simulateResiduals(fittedModel = water.mod.nb2, plot = T)

##dispersal formula checks
water.mod.nb1.disnovar <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~Microbe_treatment+Food_type,data=df.all.water.long)

anova(water.mod.nb1.disnovar,water.mod.nb1) #sig***

##removing 3 way interaction
water.mod.nb1.nofm <- update(water.mod.nb1,.~.-(1|variable:Microbe_treatment:Food_type))
anova(water.mod.nb1,water.mod.nb1.nofm)
##super sig 0.000242

water.mod.nb2.nofm <- update(water.mod.nb2,.~.-(1|variable:Microbe_treatment:Food_type))
anova(water.mod.nb2,water.mod.nb2.nofm)
##also super sig

###save conditional modes
water.mod.ranef <- as.data.frame(ranef(water.mod.nb1))
#write.csv(water.mod.ranef, "water.mod.ranef.all.csv")
#saveRDS(water.mod.nb1,file="water.mod.nb1.all.rds")
```

### Nbinom1 vs. nbinom2 things

Following this wonderful resource, on page 16

Ben Bolker, Mollie Brooks, Beth Gardner, Cleridy Lennert, Mihoko Minami, October 23, 2012, Owls example: a zero-inflated, generalized linear mixed model for count data. https://groups.nceas.ucsb.edu/non-linear-modeling/projects/owls/WRITEUP/owls.pdf

```{r}
mvtab <- ddply(df.all.water.long,
               .(variable:Microbe_treatment:Food_type),
               summarise,
               callmean=mean(value),
               callvar=var(value))

q1 <- qplot(callmean,callvar,data=mvtab)
print(q1+
        ## linear (quasi-Poisson/NB1) fit
        geom_smooth(method="lm",formula=y~x-1,colour="pink")+
        ## smooth (loess)
        geom_smooth(colour="red")+
        ## semi-quadratic (NB2/LNP)
        geom_smooth(method="lm",formula=y~I(x^2)+offset(x)-1,colour="purple")+
        ## Poisson (v=m)
        geom_abline(a=0,b=1,lty=2))
```

## Larvae - all treatments

```{r}
str(df.all.larv.long)
df.all.larv.long$rowid <- as.factor(df.all.larv.long$rowid)
df.all.larv.long$Microbe_treatment <- as.factor(df.all.larv.long$Microbe_treatment)
df.all.larv.long$Food_type <- as.factor(df.all.larv.long$Food_type)
#df.all.larv.long$food.microbes <- as.factor(df.all.larv.long$food.microbes)
df.all.larv.long$Mesocosm_id <- as.factor(df.all.larv.long$Mesocosm_id)
str(df.all.larv.long)

##binomial
larv.mod.bin <- glmmTMB(cbind(value, sum-value)~Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type)+(1|rowid),family=binomial,data=df.all.larv.long)
summary(larv.mod.bin)

##full model
larv.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),disp=~variable+Food_type+Microbe_treatment,family=nbinom1,data=df.all.larv.long)
summary(larv.mod.nb1)

larv.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),disp=~variable+Food_type+Microbe_treatment,family=nbinom1,data=df.all.larv.long)

larv.mod.nb2 <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),disp=~variable+Food_type+Microbe_treatment,family=nbinom2,data=df.all.larv.long)
summary(larv.mod.nb2)

AICtab(larv.mod.nb1,larv.mod.nb2,larv.mod.bin)

larv.mod.bin.resid <- simulateResiduals(fittedModel = larv.mod.bin, plot = T)
larv.mod.nb1.resid <- simulateResiduals(fittedModel = larv.mod.nb1, plot = T)
larv.mod.nb2.resid <- simulateResiduals(fittedModel = larv.mod.nb2, plot = T)

##dispersal formula checks
larv.mod.nb1.disnovar <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~Microbe_treatment+Food_type,data=df.all.larv.long)

anova(larv.mod.nb1.disnovar,larv.mod.nb1) #sig***

##removing 3 way interaction
larv.mod.nb1.nofm <- update(larv.mod.nb1,.~.-(1|variable:Microbe_treatment:Food_type))
summary(larv.mod.nb1.nofm)
anova(larv.mod.nb1,larv.mod.nb1.nofm) #sig **

# ##removing just food
# larv.mod.nb1.nofood <- update(larv.mod.nb1,.~.-(1|variable:Food_type))
# anova(larv.mod.nb1,larv.mod.nb1.nofood) #super sig***
# 
# ##removing just microbes
# larv.mod.nb1.nomicr <- update(larv.mod.nb1,.~.-(1|variable:Microbe_treatment))
# anova(larv.mod.nb1,larv.mod.nb1.nomicr) #super sig***

##checking out nbinom2
larv.mod.nb2.nofm <- update(larv.mod.nb2,.~.-(1|variable:Microbe_treatment:Food_type))
anova(larv.mod.nb2,larv.mod.nb2.nofm)
##sig**

##checking out binomial
larv.mod.bin.nofm <- update(larv.mod.bin,.~.-(1|variable:Microbe_treatment:Food_type))
summary(larv.mod.bin.nofm)
anova(larv.mod.bin,larv.mod.bin.nofm) #sig***

###save conditional modes
larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb1))
#larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb1))
#write.csv(larv.mod.ranef, "larv.mod.ranef.all.csv")
#saveRDS(larv.mod.nb1,file="larv.mod.nb1.all.rds")

larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb2))
write.csv(larv.mod.ranef, file="larv.mod.ranef.all.NB2.csv")

larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb1.nofm))
write.csv(larv.mod.ranef, file="larv.mod.ranef.all.no3way.csv")
# library("ggstats")
# ggcoef_multicomponents(larv.mod.nb1)
```

### Nbinom1 vs. nbinom2 things

```{r}
##note: some errors/warnings if the treatments aren't factors for some reason
mvtab <- ddply(df.all.larv.long,
               .(variable:Microbe_treatment:Food_type),
               summarise,
               callmean=mean(value),
               callvar=var(value))

q1 <- qplot(callmean,callvar,data=mvtab)
print(q1+
        ## linear (quasi-Poisson/NB1) fit
        geom_smooth(method="lm",formula=y~x-1,colour="pink")+
        ## smooth (loess)
        geom_smooth(colour="red")+
        ## semi-quadratic (NB2/LNP)
        geom_smooth(method="lm",formula=y~I(x^2)+offset(x)-1,colour="purple")+
        ## Poisson (v=m)
        geom_abline(a=0,b=1,lty=2))#+
        #ylim(0,1.75e8)+
        #xlim(0,20000)
```

## Re-doing models?

In the style of Matt's metagenome analysis

```{r}
ps.exp.all
ps.exp.all.trim <- filter_taxa(ps.exp.all, function(x) sum(x > 1) > (0.06*length(x)), TRUE)
ps.exp.all.trim #109 taxa

##before trimming
otu.exp.all <- data.frame(ps.exp.all@otu_table)
sum(otu.exp.all == 0) / (ncol(otu.exp.all) * nrow(otu.exp.all)) * 100
#95.11868%

##after trimming
otu.exp.all.trim <- data.frame(ps.exp.all.trim@otu_table)
sum(otu.exp.all.trim == 0) / (ncol(otu.exp.all.trim) * nrow(otu.exp.all.trim)) * 100
#80.11009%

##save if ready
#saveRDS(ps.exp.lar.trim,file="ps.exp.lar.trim.rds")
#saveRDS(ps.exp.all.lar.trim,file="ps.exp.all.lar.trim.rds")

#total read count colums
otu.exp.all.trim$sum <- rowSums(otu.exp.all.trim)
otu.exp.all.trim$Short_label <- row.names(otu.exp.all.trim)

#bring in the metadata again
samdf.exp.all <- data.frame(ps.exp.all.trim@sam_data)

# #join the data sets by sample name
df.exp.all.trim <- plyr::join(otu.exp.all.trim, samdf.exp.all, by="Short_label", type="left")
colnames(df.exp.all.trim)

##getting rid of some irrelevant metadat
df.exp.all.trim1 <- df.exp.all.trim %>%
  dplyr::select(-Short_label,-Longer_name,-Num_larvae,-Date_collected,-Exp_day,-Mesocosm_treatment,-Mesocosm_type,-Treatment_notes,-Notes,-Raw_reads,-Num_larvae_d00,-Num_larvae_d08,-Num_larvae_d34,-Num_larvae_d40,-Total_pupae,-Proportion_pupated,-Biomass_day30,-lib_size_clean,-is.neg)
colnames(df.exp.all.trim1)

#df.larv2 <- df.larv1[df.larv1$sum!=0,]
#df.larv2$sum==0

df.exp.all.long <- melt(df.exp.all.trim1, id.vars=c("sum", "Mesocosm_id", "Microbe_treatment","Food_type","Sample_type", "Biomass_day40"))

df.exp.all.long$rowid <- 1:nrow(df.exp.all.long)

##save
#saveRDS(df.water.long,file="df.water.long.rds")
#saveRDS(df.larv.long,file="df.larv.long.rds")
```

## New statz

```{r}
summary(mod.full.nb2 <- glmmTMB(value~offset(log(sum))+Sample_type+Food_type+Microbe_treatment+Sample_type:Food_type+(1|variable)+(1|variable:Sample_type)+(1|variable:Food_type)+(1|variable:Microbe_treatment)+(1|variable:Sample_type:Food_type), family=nbinom2, data=df.exp.all.long))

summary(mod.full.nb1 <-glmmTMB(value~offset(log(sum))+Sample_type+Food_type+Microbe_treatment+Sample_type:Food_type+(1|variable)+(1|variable:Sample_type)+(1|variable:Food_type)+(1|variable:Microbe_treatment)+(1|variable:Sample_type:Food_type), family=nbinom1, data=df.exp.all.long))

AICtab(mod.full.nb2,mod.full.nb1)

##removing 3 way interaction
mod.full.nb1.nofm <- update(mod.full.nb1,.~.-(1|variable:Sample_type:Food_type))
summary(mod.full.nb1.nofm)
anova(mod.full.nb1,mod.full.nb1.nofm) #sig ***
```

## 2-way interactions

### Just larvae first

```{r}
df.all.larv.long <- readRDS("df.all.larv.long.rds")

str(df.all.larv.long)
df.all.larv.long$rowid <- as.factor(df.all.larv.long$rowid)
df.all.larv.long$Microbe_treatment <- as.factor(df.all.larv.long$Microbe_treatment)
df.all.larv.long$Food_type <- as.factor(df.all.larv.long$Food_type)
#df.all.larv.long$food.microbes <- as.factor(df.all.larv.long$food.microbes)
df.all.larv.long$Mesocosm_id <- as.factor(df.all.larv.long$Mesocosm_id)
str(df.all.larv.long)

##binomial
#larv.mod.bin <- glmmTMB(cbind(value, sum-value)~Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type)+(1|rowid),family=binomial,data=df.all.larv.long)
#summary(larv.mod.bin)

##full model
# larv.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment+Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),disp=~variable+Food_type+Microbe_treatment,family=nbinom1,data=df.all.larv.long)
# summary(larv.mod.nb1)

larv.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment+Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type),family=nbinom1,data=df.all.larv.long)
summary(larv.mod.nb1)

larv.mod.nb2 <- glmmTMB(value~offset(log(sum))+Microbe_treatment+Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type),family=nbinom2,data=df.all.larv.long)
summary(larv.mod.nb2)

AICtab(larv.mod.nb1,larv.mod.nb2)

#larv.mod.bin.resid <- simulateResiduals(fittedModel = larv.mod.bin, plot = T)
larv.mod.nb1.resid <- simulateResiduals(fittedModel = larv.mod.nb1, plot = T)
larv.mod.nb2.resid <- simulateResiduals(fittedModel = larv.mod.nb2, plot = T)

##dispersal formula checks
#larv.mod.nb1.disnovar <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~Microbe_treatment+Food_type,data=df.all.larv.long)

#anova(larv.mod.nb1.disnovar,larv.mod.nb1) #sig***

##removing 2 way interactions
larv.mod.nb1.nomicr <- update(larv.mod.nb1,.~.-(1|variable:Microbe_treatment))
summary(larv.mod.nb1.nomicr)
anova(larv.mod.nb1,larv.mod.nb1.nomicr) #sig ***

larv.mod.nb1.nofood <- update(larv.mod.nb1,.~.-(1|variable:Food_type))
summary(larv.mod.nb1.nofood)
anova(larv.mod.nb1,larv.mod.nb1.nofood) #sig **

larv.mod.ranef.2ways <- as.data.frame(ranef(larv.mod.nb1))
larv.mod.ranef.2ways
#write.csv(larv.mod.ranef.2ways,file="larv.mod.ranef.2ways.csv")

# ##removing just food
# larv.mod.nb1.nofood <- update(larv.mod.nb1,.~.-(1|variable:Food_type))
# anova(larv.mod.nb1,larv.mod.nb1.nofood) #super sig***
# 
# ##removing just microbes
# larv.mod.nb1.nomicr <- update(larv.mod.nb1,.~.-(1|variable:Microbe_treatment))
# anova(larv.mod.nb1,larv.mod.nb1.nomicr) #super sig***

##checking out nbinom2
larv.mod.nb2.nofm <- update(larv.mod.nb2,.~.-(1|variable:Microbe_treatment:Food_type))
anova(larv.mod.nb2,larv.mod.nb2.nofm)
##sig**

##checking out binomial
larv.mod.bin.nofm <- update(larv.mod.bin,.~.-(1|variable:Microbe_treatment:Food_type))
summary(larv.mod.bin.nofm)
anova(larv.mod.bin,larv.mod.bin.nofm) #sig***

###save conditional modes
larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb1))
#larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb1))
#write.csv(larv.mod.ranef, "larv.mod.ranef.all.csv")
#saveRDS(larv.mod.nb1,file="larv.mod.nb1.all.rds")

# larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb2))
# write.csv(larv.mod.ranef, file="larv.mod.ranef.all.NB2.csv")
# 
# larv.mod.ranef <- as.data.frame(ranef(larv.mod.nb1.nofm))
# write.csv(larv.mod.ranef, file="larv.mod.ranef.all.no3way.csv")
# library("ggstats")
# ggcoef_multicomponents(larv.mod.nb1)
```

### Just waterae first

```{r}
df.all.water.long <- readRDS("df.all.water.long.rds")

str(df.all.water.long)
df.all.water.long$rowid <- as.factor(df.all.water.long$rowid)
df.all.water.long$Microbe_treatment <- as.factor(df.all.water.long$Microbe_treatment)
df.all.water.long$Food_type <- as.factor(df.all.water.long$Food_type)
#df.all.water.long$food.microbes <- as.factor(df.all.water.long$food.microbes)
df.all.water.long$Mesocosm_id <- as.factor(df.all.water.long$Mesocosm_id)
str(df.all.water.long)

##binomial
#water.mod.bin <- glmmTMB(cbind(value, sum-value)~Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type)+(1|rowid),family=binomial,data=df.all.water.long)
#summary(water.mod.bin)

##full model
# water.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment+Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),disp=~variable+Food_type+Microbe_treatment,family=nbinom1,data=df.all.water.long)
# summary(water.mod.nb1)

water.mod.nb1 <- glmmTMB(value~offset(log(sum))+Microbe_treatment+Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type),family=nbinom1,data=df.all.water.long)
summary(water.mod.nb1)

water.mod.nb2 <- glmmTMB(value~offset(log(sum))+Microbe_treatment+Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type),family=nbinom2,data=df.all.water.long)
summary(water.mod.nb2)

AICtab(water.mod.nb1,water.mod.nb2)

#water.mod.bin.resid <- simulateResiduals(fittedModel = water.mod.bin, plot = T)
water.mod.nb1.resid <- simulateResiduals(fittedModel = water.mod.nb1, plot = T)
water.mod.nb2.resid <- simulateResiduals(fittedModel = water.mod.nb2, plot = T)

##dispersal formula checks
#water.mod.nb1.disnovar <- glmmTMB(value~offset(log(sum))+Microbe_treatment*Food_type+(1|variable)+(1|variable:Microbe_treatment)+(1|variable:Food_type)+(1|variable:Microbe_treatment:Food_type),family=nbinom1,dispformula=~Microbe_treatment+Food_type,data=df.all.water.long)

#anova(water.mod.nb1.disnovar,water.mod.nb1) #sig***

##removing 2 way interactions
water.mod.nb1.nomicr <- update(water.mod.nb1,.~.-(1|variable:Microbe_treatment))
summary(water.mod.nb1.nomicr)
anova(water.mod.nb1,water.mod.nb1.nomicr) #sig ***

water.mod.nb1.nofood <- update(water.mod.nb1,.~.-(1|variable:Food_type))
summary(water.mod.nb1.nofood)
anova(water.mod.nb1,water.mod.nb1.nofood) #sig **

water.mod.ranef.2ways <- as.data.frame(ranef(water.mod.nb1))
water.mod.ranef.2ways
#write.csv(water.mod.ranef.2ways,file="water.mod.ranef.2ways.csv")

# ##removing just food
# water.mod.nb1.nofood <- update(water.mod.nb1,.~.-(1|variable:Food_type))
# anova(water.mod.nb1,water.mod.nb1.nofood) #super sig***
# 
# ##removing just microbes
# water.mod.nb1.nomicr <- update(water.mod.nb1,.~.-(1|variable:Microbe_treatment))
# anova(water.mod.nb1,water.mod.nb1.nomicr) #super sig***

##checking out nbinom2
water.mod.nb2.nofm <- update(water.mod.nb2,.~.-(1|variable:Microbe_treatment:Food_type))
anova(water.mod.nb2,water.mod.nb2.nofm)
##sig**

##checking out binomial
water.mod.bin.nofm <- update(water.mod.bin,.~.-(1|variable:Microbe_treatment:Food_type))
summary(water.mod.bin.nofm)
anova(water.mod.bin,water.mod.bin.nofm) #sig***

###save conditional modes
water.mod.ranef <- as.data.frame(ranef(water.mod.nb1))
#water.mod.ranef <- as.data.frame(ranef(water.mod.nb1))
#write.csv(water.mod.ranef, "water.mod.ranef.all.csv")
#saveRDS(water.mod.nb1,file="water.mod.nb1.all.rds")

# water.mod.ranef <- as.data.frame(ranef(water.mod.nb2))
# write.csv(water.mod.ranef, file="water.mod.ranef.all.NB2.csv")
# 
# water.mod.ranef <- as.data.frame(ranef(water.mod.nb1.nofm))
# write.csv(water.mod.ranef, file="water.mod.ranef.all.no3way.csv")
# library("ggstats")
# ggcoef_multicomponents(water.mod.nb1)
```
